### PR TITLE
Nengo3

### PR DESCRIPTION
**Pull request by atait**
**Motivation and context:**
 
Progress towards compatibility with Nengo 3.0. This makes steps towards closing #165 
 
**How long should this take to review?**
 
* Quick (less than 40 lines changed or changes are straightforward)
* To finish compatibility, longer time will be needed (see sticking points below)
 
**Types of changes:**

* Renaming and simple code changes
* Removal of python 2 compatibility features
* Extra features required in nengo 3.0.0 (e.g. `make_state`)
* Testing framework, but not quantity of tests

**Checklist:**

* [x]  I have read the **CONTRIBUTING.rst** document.
* [ ]  I have updated the documentation accordingly.
* [ ]  I have included a changelog entry.
* [ ]  I have added tests to cover my changes.
* [ ]  I have run the test suite locally and all tests passed.

## Summary
For reference, see [nengo 3.0.0 changelog](https://www.nengo.ai/nengo/changelog.html)

### Simple changes
- Stripping python 2
    - replace all `compat` components (is_iterable, is_number, etc.)
    - xrange to range
    - `np.asarray` to `np.fromiter`

- Use `make_state`
    - This is now needed before `make_step`

- Simple renaming
    - `ProbeDict` to `SignalDict`
    - `allclose` to `signals_allclose`

### Testing framework compatible with updated `nengo-bones` strategy
    - get `pytest-allclose`, `pytest_rng`, and `pytest_plt` from new locations
    - disabled logging or substituted `capsys` from pytest
        - TODO: use pytest to restore desired logging capabilities
    - removed `--simulator` and `--reference-simulator` from tox.ini
    - The pytest `addopts` from `pytest_nengo.py` are now being discovered automatically, so `-p nengo.tests.options` is unnecessary. Tip: be careful that `pytest_nengo.py` is not double-installed in the pytest scope (look for this file on your PYTHONPATH, not just in your installed packages)

#### Special attention: State-space filter
This [PR](https://github.com/nengo/nengo/pull/1535) is an upgrade in efficiency and accuracy from @hunse and requires significant changes to the internal representation of `LinearFilter`. Changes:

- The v2.8 `make_step` does some preprocessing of num and den, so that code needed to be copied over to Nengo OCL

- Nengo OCL is expecting to get numerators and denominators, so `ss2tf` is used to recover them from the new representation. It also wanted types `NoDen` and `Simple` instead of the new `NoX`, `NoD`, etc. Status:
    - Behavior on Alpha response is patchy because it is 2-pole, so
    - I think the current solution to this will fail in the general case
    - __quick solution__ is to debug my array handling `Simulator._plan_linearfilter`
    - __good solution__ would be to write a new `plan_linearfilter` for state-space calculations.

- Subtle change to `make_step` interface:

See changes to `test_clra_nonlinearities.py:test_linearfilter` for an example. The OCL test there works vs. the 2.8 version, but I had to *change the reference simulation* in the following ways to pass the test in 3.0.0
    - `size_in=(n,1)` is now `size_in=(n,)`
    - to get an input ndarray `x` from RaggedArray `X` now involves a transpose
    - In other words, the nengo 3.0 API might have swapped some dimensions on input and output, and the OCL API should change with it: TODO
    - __good solution__ is to 1) investigate why the nengo interface might have changed, and 2) update the nengo OCL `plan_linearfilter` interface, instead of changing the reference simulator in this test

#### Special attention: Planner
Comparing the `greedy_planner.op_groups` attribute from OCL to the `Simulator.step_order` from nengo core gives different results. There is no longer an `ElementwiseInc`; there are new operations related to PES; there is some reordering of the copys. This makes a test of `assert len(sim.op_groups == 9)` fail, but it's unclear whether that matters.

### CODE REVIEW

**Comment by drasmuss** [comment]
Hi @atait, just wanted to let you know that we are planning on taking a look at this and merging it in, just haven't had the cycles! But hopefully we should get to it within the next week or two.

**Comment by hunse** [comment]
Ok, I've got this merged into the `nengo3` branch, and added some fixes so that all the Nengo 3 tests pass (except the ones for features that we don't support).

Currently, this only supports Nengo 3.0.0 exactly, not Nengo `master` branch. I want to look into how much work it would be to get this working with Nengo `master`.

I haven't added support for the new transforms in Nengo 3 (namely `Sparse` and `Convolution`). This repo has some existing convolution code, so it would be nice to get that working with the `Convolution` transform, but that's another day's work (both literally and figuratively).

**Comment by atait** [comment]
One option is handling `nengo` `Sparse` transforms and letting them go on CPU as-is. That way it is at least supported. I have found that I'm either dealing with always very dense (nonzero frac ~ 1) or always very sparse (nonzero frac ~ 1e-4) matrices, so the latter might actually do better on a CPU anyways.

Just based on poking around the internet, the "good solution" seems like adding [clSPARSE](https://github.com/clMathLibraries/clSPARSE) support into PyOpenCL – an additional layer of development! Meanwhile, cuSPARSE is supported by PyCUDA, the CSR aspect of which is making its way [into Tensorflow](https://github.com/tensorflow/community/pull/246) (thanks for the link @drasmuss!), and hopefully Nengo DL.

So I would vote for the CPU fallback option in Nengo OCL and dealing with the latest HPC developments in Nengo DL. I can't really comment on `Convolution` from my experience.

**Comment by atait** [comment]
I've been using the nengo3 branch this week, and it's working great for large simulations on Nvidia and AMD cards. I've tried out a fallback to sparse matrices by wrapping the `scipy.sparse.csr_matrix.dot` function within a `Process`. The OCL builder is forced to fall back and run the process as a python function on CPU (see below)

The sparse step on CPU takes about 100x longer than the rest of the GPU steps. The dot is extremely sparse, so in an ideal world with free data movement, it would be the other way around. Unfortunately, that negates the reasons for using Nengo OCL. 

For neural networks though, you don't need a whole sparse library. You really just need a CSR matrix-vector dot product. Writing a kernel that stores and multiplies one row seems reasonable. Python/C-ish pseudocode:

Declaration:
```cpp
W_i = list(tuple(j, w_ij))
```
where `W_i` is weight vector of neuron `i`. It's different on each GPU kernel, and stored in GPU. length(W_i) is ∑nonzero(w_ij). 

Program:
```cpp
J_i = 0
for j, wj in W_i:
    J_i += wj * S[j]
// Existing neuron update using state V and input J to get spiked outS
outS = lif_or_some_other_nonlinearity(J_i, V)
```
where `S` is the spike vector. The elements of `S` are populated in parallel, and then the whole vector has to broadcast to every kernel, as opposed to (currently) sending one element of J to each kernel. 

I'm confused about how to combine this element-wise up with broadcasting down. Does this seem reasonable enough to code manually? I am just not very familiar with GPU language and data handling.

Implementation of fallback I used:
```python
class SparseDotPy(Process):
    def __init__(self, spmat):
        self.step = lambda t, x: spmat.dot(x)
        super().__init__(default_size_in=spmat.shape[1], 
                         default_size_out=spmat.shape[0])
    def make_step(self, shape_in, shape_out, dt, rng, state):
        return self.step
```

**Comment by hunse** [comment]
All the code here has been merged into the `nengo3` branch, so I'm closing this and opening a new PR to merge that into master. Thanks again for all your great work, @atait!

For sparse transforms, I've just got them on the host for now. To do them on the GPU, we need to figure out some data format to store them. The most formal way to do this would be to make a version of `RaggedArray` for sparse matrices. A quicker hack would be to just map individual matrices to PyOpenCL arrays for the data and the indices. Then we'd need to write a kernel to do this, which doesn't have to be complicated, but may be able to benefit from some optimizations in data loading (though it's much harder to optimize because data access is so dependent on the indices). Anyway, that will be future work.

**Comment by hunse** [comment]
New PR is #172

**Comment by hunse** [comment]
There are some OpenCL kernels for sparse matrix multiplication here (under MIT license), that we might want to use: https://lists.tiker.net/hyperkitty/list/pyopencl@tiker.net/thread/ZV6IZTQG4DVU5SMRC5MAS7ON5GAG3UUD/

**Comment by atait** [comment]
Since Numpy's SparseMatrix is frozen, Nengo OCL wouldn't have to support repeated transfers to device. What if you declare the data at render time and just forget about representing a new data structure? It would be nice if it weren't frozen, but that's an issue for nengo core.

Pseudocode [wrong types and scopes, probably wrong expansion of basecsr attributes within render, missing striding calculations]
```c
const int N = ${basecsr.getnnz()} // number of nonzero elements
const float* Val = ${basecsr.data} // contains the value of the non-zero elements
const int* RowPtr = ${basecsr.indptr} // contains the row-index range of the non-zero elements
const int* Col = ${basecsr.indices} // contains the column-index of the non-zero elements

int i = get_global_index(0) // row of this worker
Y[i] = 0
for (k = RowPtr[i]; k < RowPtr[i+1]; k = k + 1)
    Y[i] += Val[k] * X[Col[k]]; 
```
and render that after 
```python
textconf.update(basecsr=nengo_sparse_transform.init.allocate())
```
